types/share_splitting.go

```diff
@@ -51,63 +102,110 @@ func splitMessage(rawData []byte, nid namespace.ID) NamespacedShares {
-	}
-	return shares
+type ContiguousShareWriter struct {
+	shares       []NamespacedShare
+	pendingShare NamespacedShare
+	namespace    namespace.ID
+}
+
+func NewContiguousShareWriter(ns namespace.ID) *ContiguousShareWriter {
+	pendingShare := NamespacedShare{ID: ns, Share: make([]byte, 0, consts.ShareSize)}
+	pendingShare.Share = append(pendingShare.Share, ns...)
+	return &ContiguousShareWriter{pendingShare: pendingShare, namespace: ns}
 }
 
-// getNextChunk gets the next chunk for contiguous shares
-// Precondition: none of the slices in rawDatas is zero-length
-// This precondition should always hold at this point since zero-length txs are simply invalid.
-func getNextChunk(rawDatas [][]byte, outerIndex int, innerIndex int, width int) ([]byte, int, int, int) {
-	rawData := make([]byte, 0, width)
-	startIndex := 0
-	firstBytesToFetch := 0
+// Write adds the delimited data to the underlying contiguous shares
+func (csw *ContiguousShareWriter) Write(rawData []byte) {
+	// if this is the first time writing to a pending share, we must add the
+	// reserved bytes
+	if len(csw.pendingShare.Share) == consts.NamespaceSize {
+		csw.pendingShare.Share = append(csw.pendingShare.Share, 0)
+	}
 
-	curIndex := 0
-	for curIndex < width && outerIndex < len(rawDatas) {
-		bytesToFetch := min(len(rawDatas[outerIndex])-innerIndex, width-curIndex)
-		if bytesToFetch == 0 {
-			panic("zero-length contiguous share data is invalid")
-		}
-		if curIndex == 0 {
-			firstBytesToFetch = bytesToFetch
-		}
-		// If we've already placed some data in this chunk, that means
-		// a new data segment begins
-		if curIndex != 0 {
-			// Offset by the fixed reserved bytes at the beginning of the share
-			startIndex = firstBytesToFetch + consts.NamespaceSize + consts.ShareReservedBytes
+	txCursor := len(rawData)
+	for txCursor != 0 {
+		// find the len left in the pending share
+		pendingLeft := consts.ShareSize - len(csw.pendingShare.Share)
+
+		// if we can simply add the tx to the share without creating a new
+		// pending share, do so and return
+		if len(rawData) <= pendingLeft {
+			csw.pendingShare.Share = append(csw.pendingShare.Share, rawData...)
+			break
 		}
-		rawData = append(rawData, rawDatas[outerIndex][innerIndex:innerIndex+bytesToFetch]...)
-		innerIndex += bytesToFetch
-		if innerIndex >= len(rawDatas[outerIndex]) {
-			innerIndex = 0
-			outerIndex++
+
+		// if we can only add a portion of the transaction to the pending share,
+		// then we add it and add the pending share to the finalized shares.
+		chunk := rawData[:pendingLeft]
+		csw.pendingShare.Share = append(csw.pendingShare.Share, chunk...)
+		csw.stackPending()
+
+		// update the cursor
+		rawData = rawData[pendingLeft:]
+		txCursor = len(rawData)
+
+		// add the share reserved bytes to the new pending share
+		pendingCursor := len(rawData) + consts.NamespaceSize + consts.ShareReservedBytes
+		var reservedByte byte
+		if pendingCursor >= consts.ShareSize {
+			reservedByte = byte(0)
```

Comment:
Can we document why this must be a 0 byte?

Reply:
yeah, that's a good idea https://github.com/celestiaorg/celestia-core/pull/637/commits/8e541035ba29b8eb957d8d213273b240586a259c